ambry-api/src/main/java/com.github.ambry/router/ByteRange.java

```diff
@@ -17,208 +17,110 @@
 /**
  * Represents a byte range for performing ranged get requests.
  */
-public class ByteRange {
-  private static final long UNDEFINED_OFFSET = -1;
-
-  private final ByteRangeType type;
-  private final long startOffset;
-  private final long endOffset;
+public abstract class ByteRange {
 
   /**
    * Construct a range from a start offset to an end offset.
    * @param startOffset the (inclusive) start byte offset.
    * @param endOffset the (inclusive) end byte offset.
    * @return A {@link ByteRange} with the specified offsets.
-   * @throws IllegalArgumentException
+   * @throws IllegalArgumentException if the start offset is less than 0, or the end offset is less than the start.
    */
   public static ByteRange fromOffsetRange(long startOffset, long endOffset) {
     if (startOffset < 0 || endOffset < startOffset) {
       throw new IllegalArgumentException(
           "Invalid range offsets provided for ByteRange; startOffset=" + startOffset + ", endOffset=" + endOffset);
     }
-    return new ByteRange(startOffset, endOffset, ByteRangeType.OFFSET_RANGE);
+    return new OffsetRange(startOffset, endOffset);
```

Comment:
An abstract class shouldn't create instances of its subclasses.  Could test static constructors be moved to a factory class eg. ByteRangeFactory?

Reply:
abstract class instantiating subclasses does feel a bit odd. I do like the idea of avoiding the switch logic though. Is there a better way here?

Reply:
Would deprecation be an acceptable compromise? I could move the impls and static methods into a `ByteRanges` util class:

```java
public abstract class ByteRange {
  // use ByteRanges.fromStartOffset instead
  @Deprecated
  static ByteRange fromStartOffset(long start, long end) {
    return ByteRanges.fromStartOffset(start, end);
  }
  ...
  abstract ByteRange toResolvedByteRange(long totalSize);
  ...
  long getStartOffset() {
    throw new UnsupportedOperationException();
  }
```

```java
public class ByteRanges {
  public ByteRange fromStartOffset(long start, long end) {
    return new ClosedRange(start, end);
  }
  ...
  private static class ClosedRanged extends ByteRange {
    ...
  }
}
```

Reply:
LGTM